config/base/derive/src/proxy.rs

```diff
@@ -0,0 +1,217 @@
+                quote! { self.#ident.write().await }
+            } else {
+                quote! { self.#ident }
+            }
+        })
+        .collect::<Vec<_>>();
+
+    let as_str = field_attrs
+        .iter()
+        .map(|attrs| {
+            attrs
+                .iter()
+                .any(|attr| attr.parse_args::<config::SerdeAsStr>().is_ok())
+        })
+        .collect::<Vec<_>>();
+
+    println!("GETTING PROXY LOADENV");
+    // dbg!(&field_ty);
+    // dbg!(&field_idents);
+    let proxy_ty = field_ty
+        .iter()
+        .map(|ty| {
+            let new_ty: Type = parse_quote! {
+                // #(#attrs)*
+                Option<#ty>
+            };
+            new_ty
+        })
+        .collect::<Vec<_>>();
+
+    // let load_env_fn = config::impl_load_env(
+    //     &field_idents,
+    //     &inner,
+    //     &lvalue_write,
+    //     &as_str,
+    //     &proxy_ty,
+    //     &field_environment,
+    // );
+
+    let proxy_fields = gen_proxy_struct_fields(&field_idents, &field_attrs, &field_ty);
+    let build_fn = impl_build(&field_idents, &field_attrs);
+
+    quote! {
+        // #[derive(Debug, Clone, Serialize, Deserialize)]
+        // // #(#attrs)*
+        // #vis struct #proxy_name #generics {
+        //     #proxy_fields
+        // }
+        // impl iroha_config_base::proxy::Combine for #proxy_name {
+        //     type Target = #parent_name;
+        //     #load_env_fn
+        //     #build_fn
+
+        // }
+    }
+    .into()
+}
+
+fn impl_build(field_idents: &[&Ident], field_attrs: &[&[Attribute]]) -> proc_macro2::TokenStream {
+    let checked_fields = gen_none_fields_check(field_idents, field_attrs);
+
+    quote! {
+        fn build(self) -> Result<Self::Target, iroha_config_base::derive::Error> {
+            Ok(Self::Target {
+                #checked_fields
+            })
+        }
+    }
+}
+
+fn gen_proxy_struct_fields(
+    field_idents: &[&Ident],
+    field_attrs: &[&[Attribute]],
+    field_ty: &[Type],
+) -> proc_macro2::TokenStream {
+    let combined_fields = field_idents
+        .iter()
+        .zip(field_attrs.iter())
+        .zip(field_ty.iter())
+        .map(|((ident, attrs), ty)| {
+            if utils::is_option_type(ty) {
+                quote! { #ident: #ty }
+            } else {
+                quote! {
+                    // #(#attrs)*
+                    #ident: Option<#ty>
+                }
+            }
+        });
+    quote! {
+        #(#combined_fields),*
+    }
+}
+
+/// Helper function for checking inner
+fn gen_none_fields_check(
+    field_idents: &[&Ident],
+    field_attrs: &[&[Attribute]],
+) -> proc_macro2::TokenStream {
+    let checked_fields = field_idents
+        .iter()
+        .zip(field_attrs.iter())
+        .map(|(ident, attrs)| {
+            quote! {
+                // #(#attrs)*
+                #ident: self.#ident.ok_or(iroha_config_base::derive::Error::ProxyError(ident.to_string()))?
+            }
```

Comment:
This is how fields would be checked for presence at build time.

Reply:
Aha, so this is the case against option flattening. So no special handling is needed, just as you mentioned [here](https://github.com/hyperledger/iroha/pull/2583#discussion_r937729839). Scrapping that part for sure then. Thanks for a good catch!